waiter/src/waiter/auth/authentication.clj

```diff
@@ -201,3 +228,110 @@
+
+(defn process-auth-keep-alive-request
+  "Handler to eagerly trigger authentication workflow even if cookie has not yet expired.
+   This allows clients to pre-emptively refresh credentials before they expire.
+   Presence of done parameter is used to avoid infinite auth redirect loops and will return 204 No Content.
+   Invalid offset parameters result in a 400 error.
+   Missing offset, soon to expire cookie (based on offset) or invalid auth cookie will trigger the authentication workflow.
+   If cookie is expected to be live longer than offset value, return 204 No Content."
+  [password auth-handler {:keys [headers query-string] :as request}]
+  (let [{:strs [done offset]} (-> request ru/query-params-request :query-params)
+        offset-parsed (utils/parse-int offset)
+        decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ cookie-metadata] decoded-auth-cookie
+        current-epoch-time (tc/to-epoch (t/now))
+        {:keys [expires-at]} cookie-metadata
+        ;; handle legacy cookies which will not have this value set
+        expires-at (or expires-at current-epoch-time)
+        waiter-token (some-> request utils/request->host utils/authority->host)]
+    (log/info auth-principal "cookie expires at" expires-at "offset is" offset-parsed)
+    (cond
+      ;; avoid infinite redirect loop
+      done
+      (-> {:status http-204-no-content}
+        (utils/attach-waiter-source)
+        (assoc :waiter/token waiter-token))
+
+      ;; offset parameter provided, but cannot be parsed
+      (and offset (nil? offset-parsed))
+      (-> {:message "Unable to parse offset parameter"
+           :parameter {:offset offset}}
+        (utils/clj->json-response :status http-400-bad-request)
+        (assoc :waiter/token waiter-token))
+
+      ;; offset parameter must be positive when provided
+      (and offset-parsed (not (pos? offset-parsed)))
+      (-> {:message "Invalid offset parameter"
+           :parameter {:offset offset-parsed}}
+        (utils/clj->json-response :status http-400-bad-request)
+        (assoc :waiter/token waiter-token))
+
+      ;; trigger auth workflow if
+      ;; - offset query parameter is missing;
+      ;; - cookie has already expired; or
+      ;; - including offset time will cause the cookie to expire
+      (or (nil? offset)
+          (not (decoded-auth-valid? decoded-auth-cookie))
+          (>= (+ current-epoch-time offset-parsed) expires-at))
+      (do
+        (log/info "initiating authentication flow for keep-alive")
+        (-> request
+          ;; ensure loop back to this endpoint terminates
+          (assoc :query-string (str query-string (when query-string "&") auth-keep-alive-done-parameter))
+          ;; remove existing auth cookies
+          (update-in [:headers "cookie"] remove-auth-cookie)
+          ;; trigger auth as if it is a proxy request
+          (assoc :waiter-api-call? false)
```

Comment:
This certainly feels like an API call.  Please document why this should not be considered API call.

Reply:
Added comment, OIDC and JWT auth behavior may be different for Proxy and Waiter api requests